projects/extension/src/types/index.ts

```diff
@@ -32,24 +32,34 @@ export type ExtensionMessageType = 'error' | 'rpc';
 
 export interface ExtensionMessage {
   type: ExtensionMessageType;
-  payload: string;
+  payload: string; // JSON encoded RPC response or an error message
 }
 
 // Messages that come from the app
 export type AppMessageType = 'associate' | 'rpc';
 
+/* The inner message that was received by the content script to be sent on to
+ * the extension background.
+ */
 export interface AppMessage {
   type: AppMessageType;
-  payload: string; // smoldot name / json / message_id / subscription_id
+  payload: string; // smoldot name or JSON encoded RPC message
   subscription?: boolean;
 }
 
-export interface Message {
+/* A message that is sent by the `ExtenionProvider` calling `window.postMessage`
+ * that is received by the content script.
+ *
+ * The `message` property will either be an `AppMessage` for sending through the
+ * port to the background script or the string 'disconnect' to tell the content
+ * script to disconnect the port.
+ */
+export interface ExtensionProviderMessage {
   data: {
     appName: string;
     chainName: string;
     origin: string;
-    message: AppMessage
+    message: AppMessage | 'disconnect';
```

Comment:
This confuses the types a bit. You assign to message either a json or a string. Its best if 'disconnect' is part of AppMessageType

Reply:
The AppMessage type is what gets passed unchanged to the extension.  Putting it there would make the logic for forwarding messages more complicated.  This information is to be used by the router only and not by the background. I could put it on the `data` property of `ExtensionProviderMessage` as a sibling of `message`?

Reply:
I.E. I could add an `action` property which would be one of `'forward' | 'disconnect'`.  This would describe the action that the router should take when processing this message.  The `message` will be empty for 'disconnect' and it should disconnect the port and it will be an `AppMessage` for 'forward' when it is either an associate or rpc message and the router should take the action of forwarding that message.  Does that cover your concerns adequately?

Reply:
Afaiu the 'disconnect' is used as action/status rather than message in this case. If that is it then  a "flag" or different type of structure should return. if this is supposed to be the message then previous comment applies

Reply:
Yup I agree :+1:

Reply:
> I.E. I could add an `action` property which would be one of `'forward' | 'disconnect'`. This would describe the action that the router should take when processing this message. The `message` will be empty for 'disconnect' and it should disconnect the port and it will be an `AppMessage` for 'forward' when it is either an associate or rpc message and the router should take the action of forwarding that message. Does that cover your concerns adequately?

I implemented this in the latest commit